### PR TITLE
Object store/ connection check on startup

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Updated startup checks for cloud providers to make sure a connection is made to the bucket/container otherwise the startup is exited. Also, updated the filesystem startup to check if a basedirectory is writeable. If not then startup is exited as well.

**To test cloud connection:**
1) In the `irida.conf` file change the bucket name for aws or the container name for azure. Also, change the `irida.storage.type` to either aws or azure
2) Startup IRIDA
3) IRIDA should not startup and there should be an exception in the console advising you to check the credentials and that the bucket/container exist

**To test local file system directory writable check:**
1) Make one of the output directories in `irida.conf` non writable
2) Startup IRIDA
3) IRIDA should not startup and there should be an exception in the console advising you that the directory is not writable and to check the directory permissions

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

~* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.~
~* [ ] Tests added (or description of how to test) for any new features.~
~* [ ] User documentation updated for UI or technical changes.~

### CODE REVIEW

**Review by tom114 — CHANGES_REQUESTED** [review]
Hey thanks for the changes.  A few questions and fixes below.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/repository/IridaApiFilesystemRepositoryConfig.java [hunk 1]
```diff
@@ -91,15 +98,37 @@ public Path assemblyFileBaseDirectory() throws IOException {
 		return getExistingPathOrThrow(assemblyFileBaseDirectory);
 	}
 
-	//FIXME Update the code to check if the connection to a cloud provider
-	// is successful or not as well as check if path is writeable for local file storage
 	private Path getExistingPathOrThrow(String directory) {
 		Path baseDirectory = Paths.get(directory);
-		if (!Files.exists(baseDirectory) && storageType.equals("local")) {
+		if (!Files.exists(baseDirectory) && storageType.equalsIgnoreCase(StorageType.LOCAL.toString())) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/repository/IridaApiFilesystemRepositoryConfig.java]: By the looks of it we can't instantiate an enum as it is just static values. I added a fromString method which we can use to retrieve the storagetype before doing a comparison. Update in ad89bd2c21696c879bec53167ab4412a78fa5003

File: src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java [hunk 2]
```diff
@@ -321,14 +322,23 @@ private String getClasspathResourceBasename(Pattern pattern, Resource x) {
 	@Bean(name = "iridaFileStorageUtility")
 	public IridaFileStorageUtility iridaFileStorageService() {
 		IridaFileStorageUtility iridaFileStorageUtility;
-		if (storageType.equalsIgnoreCase("aws")) {
+		if (storageType.equalsIgnoreCase(StorageType.AWS.toString())) {
 			iridaFileStorageUtility = new IridaFileStorageAwsUtilityImpl(awsBucketName, awsBucketRegion, awsAccessKey,
 					awsSecretKey);
-		} else if(storageType.equalsIgnoreCase("azure")) {
+		} else if (storageType.equalsIgnoreCase(StorageType.AZURE.toString())) {
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java]: By the looks of it we can't instantiate an enum as it is just static values. I added a fromString method which we can use to retrieve the storagetype before doing a comparison. Update in ad89bd2

File: src/main/java/ca/corefacility/bioinformatics/irida/config/repository/IridaApiFilesystemRepositoryConfig.java [hunk 3]
```diff
@@ -91,15 +98,37 @@ public Path assemblyFileBaseDirectory() throws IOException {
...
+					if(!directoryWriteable) {
+						// Log the error and exit so startup does not continue
+						logger.error("Cannot continue startup; base directory " + baseDirectory + " does not have write access! Please check directory permissions.");
+						System.exit(1);
+					}
+				} catch (IOException e) {
+					logger.error("Unable to create temporary file. Please check directory permissions", e);
+					System.exit(1);
+				}
+			}
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/repository/IridaApiFilesystemRepositoryConfig.java]: Putting it in the checkConnectivity method wouldn't work as we use it to check the connection to the cloud providers. I created a new method in iridafilestorageutility to check for write access. Added in ad89bd2

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java [hunk 4]
```diff
@@ -123,4 +123,14 @@
...
+
+	/**
+	 * Check the connectivity to the file storage
+	 * mechanism. For cloud services we must make an
+	 * api call to check if containers/buckets exist.
+	 *
+	 * @return Connection valid or not
+	 * @throws IllegalStateException if connection failed
+	 */
+	public boolean checkConnectivity() throws IllegalStateException;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageUtility.java]: Updated to use StorageException in ad89bd2

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java [hunk 5]
```diff
@@ -220,4 +220,11 @@ public String getFileExtension(List<? extends SequencingObject> sequencingObject
 		}
 		return bytes;
 	}
+
+	@Override
+	public boolean checkConnectivity() {
+		// Since it's the local filesystem we just return true
+		return true;
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java:227]: Now that we have the `checkWriteAccess` methods up (and especially if we beef them up on the cloud platforms) this method is pretty much a duplicate.  Do we want to refactor this method out and just have the `checkWriteAccess` methods handle things?

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java [hunk 6]
```diff
@@ -220,4 +220,48 @@ public String getFileExtension(List<? extends SequencingObject> sequencingObject
...
+				// Check if basedirectory path is writeable by creating a temp file and then removing it
+				Path tempFile = Files.createTempFile(baseDirectory, "", "");
+				// Check if directory is writeable
+				boolean directoryWriteable = Files.isWritable(tempFile);
+
+				try {
+					// Cleanup the temp file created in the directory
+					Files.delete(tempFile);
+				} catch (IOException e) {
+					logger.error("An I/O error occurred while attempting to remove temp file ", e);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java:246]: This is probably not an error we just want to swallow.  This should prevent the server from starting.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java [hunk 7]
```diff
@@ -220,4 +220,48 @@ public String getFileExtension(List<? extends SequencingObject> sequencingObject
...
+				}
+
+				if (!directoryWriteable) {
+					// Log the error and exit so startup does not continue
+					logger.error("Cannot continue startup; base directory " + baseDirectory
+							+ " does not have write access! Please check directory permissions.");
+					System.exit(1);
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java:253]: I don't think we want to directly call system.exit in any of the cases in here.  It should more likely throw an exception that doesn't get caught.  That way the servlet container (tomcat) decides how to handle things.  Can the `System.exit` lines get refactored out in all the cases and just replaced with a throw (maybe `StorageException`).

**Review by tom114 — CHANGES_REQUESTED** [review]
A few more comments here.  I really think we should get off the `System.exit(1)` things.  It's kinda undefined how that would be handled when we're running in something like Tomcat.  Instead throw an exception that will get uncaught so Tomcat can display the error and handle it how it sees fit.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java [hunk 8]
```diff
@@ -321,18 +323,33 @@ private String getClasspathResourceBasename(Pattern pattern, Resource x) {
 	@Bean(name = "iridaFileStorageUtility")
 	public IridaFileStorageUtility iridaFileStorageService() {
 		IridaFileStorageUtility iridaFileStorageUtility;
-		if (storageType.equalsIgnoreCase("aws")) {
+		StorageType st = StorageType.fromString(storageType);
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java]: Injected in b47ef804ace2d29d02bac15593fc7f7e5129e205

File: src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java [hunk 9]
```diff
@@ -321,18 +323,33 @@ private String getClasspathResourceBasename(Pattern pattern, Resource x) {
...
+		} catch (StorageException e) {
+			// Log the error and exit so startup does not continue
+			logger.error("Unable to start up IRIDA! ", e);
+			System.exit(1);
+		}
 		IridaFiles.setIridaFileStorageUtility(iridaFileStorageUtility);
 		return iridaFileStorageUtility;
 	}
 
+	@Bean(name= "storageType")
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/config/services/IridaApiServicesConfig.java]: Fixed in b47ef804ace2d29d02bac15593fc7f7e5129e205

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAwsUtilityImpl.java [hunk 10]
```diff
@@ -308,4 +308,25 @@ public String getFileExtension(List<? extends SequencingObject> sequencingObject
...
+			}
+		} catch (AmazonServiceException e) {
+			throw new StorageException(
+					"Unable to connect to aws s3 bucket. Please check that your credentials are valid and that the bucket " + bucketName + " exists.");
+		}
+	}
+
+	@Override
+	public boolean checkWriteAccess(Path baseDirectory) {
+		return true;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAwsUtilityImpl.java]: That looks like the way to do it. Until we get access to aws I can't really test it. Updated in b47ef804ace2d29d02bac15593fc7f7e5129e205

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java [hunk 11]
```diff
@@ -305,4 +305,29 @@ private String getAzureFileAbsolutePath(Path file) {
...
+				connection was not successful to the azure container
+			 */
+			throw new StorageException(
+					"Unable to connect to azure container. Please check that your credentials are valid and that the container " + containerName + " exists.");
+		}
+	}
+
+	@Override
+	public boolean checkWriteAccess(Path baseDirectory) {
+		return true;
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageAzureUtilityImpl.java]: Updated in b47ef804ace2d29d02bac15593fc7f7e5129e205. We ended up having to create a temp file, upload it to azure, and delete it. This ensures we have both read and write access.

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java (hunk 6 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java]: Updated to throw exception in b47ef804ace2d29d02bac15593fc7f7e5129e205

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java (hunk 7 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java]: Refactored to use StorageException in b47ef804ace2d29d02bac15593fc7f7e5129e205

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/filesystem/IridaFileStorageLocalUtilityImpl.java:260]: Yup good idea. Removed the `checkConnectivity` method and updated `checkWriteAccess` method to check for read/write access for cloud storage.

**Review by tom114 — APPROVED** [review]

**Pull request merged**